0007-python3.md

```diff
@@ -0,0 +1,97 @@
+  Santiago Rodríguez Collazo, Miguel Angel Medinilla Luque, José Raddaoui Marín,
+  Cole Maclean, Ross Spencer, Douglas Cerna.
+* Date: 2019-07-22
+
+## Context and problem statement
+
+Python 2 will not be supported by the Python development team past January 1,
+2020\. Operating system vendors will provide bugfix support for some number of
+years, depending on the OS:
+
+* [Ubuntu 16.04 through 2021](https://wiki.ubuntu.com/Releases)
+* [Ubuntu 18.04 through 2023](https://wiki.ubuntu.com/Releases)
+* [RHEL/CentOS 7 through 2024](https://wiki.centos.org/About/Product)
+
+Archivematica is not fully compatible with Python 3 yet. Archivematica and all
+of its required components will eventually need to run in a Python 3
+environment, especially as downstream dependencies begin to drop support for
+Python 2 (i.e.
+[Django](https://docs.djangoproject.com/en/2.2/faq/install/#what-python-version-should-i-use-with-django)).
+
+Components:
+
+* [am/compose](https://github.com/artefactual-labs/am)
+* [archivematica-storage-service](https://github.com/artefactual/archivematica-storage-service)
+* [MCPServer](https://github.com/artefactual/archivematica/tree/a889605d8c97e114f8c0bc707d6a371030fb5c0b/src/MCPServer)
+* [MCPClient](https://github.com/artefactual/archivematica/tree/a889605d8c97e114f8c0bc707d6a371030fb5c0b/src/MCPClient)
+* [archivematica-common](https://github.com/artefactual/archivematica/tree/a889605d8c97e114f8c0bc707d6a371030fb5c0b/src/archivematicaCommon)
+* [Dashboard](https://github.com/artefactual/archivematica/tree/a889605d8c97e114f8c0bc707d6a371030fb5c0b/src/dashboard)
+* [mets-reader-writer](https://github.com/artefactual-labs/mets-reader-writer)
+* [ammcpc](https://github.com/artefactual-labs/ammcpc)
+* [agentarchives](https://github.com/artefactual-labs/agentarchives)
+* [fixity](https://github.com/artefactual/fixity)
+* [automation-tools](https://github.com/artefactual/automation-tools)
+* [archivematica-acceptance-tests](https://github.com/artefactual-labs/archivematica-acceptance-tests)
+* [amclient](https://github.com/artefactual-labs/amclient)
+
+## Decision drivers <!-- optional -->
+
+* Archivematica is built on Python 2
+* Python 2 will no longer be supported after January 1, 2020
+* Archivematica should be refactored to use Python 3.5 (at minimum)
+
+## Considered options
+
+1. Refactor to Python 3.5
+
+## Decision outcome
+
+Chosen option: 1, because refactoring Archivematica and all required components
+to Python 3 is the only option to ensure that Archivematica continues to be a
+safe, stable environment.
+
+Python 3.5 was chosen for the following reasons:
+
+* [Python 3.4](https://www.python.org/dev/peps/pep-0429/) has already been
+  retired
+* [Python 3.5](https://www.python.org/dev/peps/pep-0478/) is the oldest version
+  supported
+* Ubuntu 16.04 ships Python 3.5
+* Ubuntu 18.04 ships Python 3.6
+* For CentOS 7, The Software Collections Repository provides packages for
+  Python 3.5 and up
```

Comment:
I have a few thoughts here, but don't know how to write them clearly. I'll try:

I feel like the reason needs to be stronger if this is something we can control. If we have the opportunity to control installing 3.7 on all of our platforms, then I'd do that, because that will be the minimum version we have to support for what? the next year? two years? 

The official line for `3.5` is:[ _"Python 3.5 has now entered "security fixes only" mode, and as such the only changes since Python 3.5.4 are security fixes. Also, Python 3.5.6 has only been released in source code form; no more official binary installers will be produced."_](https://www.python.org/downloads/release/python-356/). For `3.6` [ _"Python 3.7 is now released and is the latest feature release of Python 3. Get the latest release of 3.7.x here. We plan to continue to provide bug-fix releases for 3.6.x though the end of 2018 and security fixes through 2021."_](https://www.python.org/downloads/release/python-360/).

If we need to support a previous Python at all, then one version back seems like it would be enough, and we can do the work needed in Ubuntu at least to install a newer version. RHEL8 [has 3.6](https://developers.redhat.com/blog/2018/11/14/python-in-rhel-8/), and so we can expect the same from CentOS8 I imagine. This seems likely to appear soon. (it looks like it is possible on CentOS7)

Python `3.8` [might drop in October](https://www.python.org/dev/peps/pep-0569/). 

It is a maintenance cost. The unknown is how much that cost is or how much it impacts developer efficacy. 

I feel like a good way to mitigating an over-promise (in this case, going too many versions back), is to say: _"we will support the minimum possible version of Python it is to install on our recommended OS"_ (RHEL?) or maybe, _"we will support the last two feature-releases of Python 3"_. I'm not sure, but just something that means we're not having to look too far back. If it is still Python 3.5 then we can still indicate it isn't set in stone, we have flexibility to move away.